smsd/core.c

```diff
@@ -2196,6 +2196,11 @@ GSM_Error SMSD_MainLoop(GSM_SMSDConfig *Config, gboolean exit_on_failure, int ma
 			if (error == ERR_EMPTY) {
 				lastnothingsent = lastloop;
 			}
+			if (error == ERR_UNKNOWN) {
+				/* Huawei devices may return UNKNOWN errors. A soft reset may help to recover */
+				SMSD_LogError(DEBUG_INFO, Config, "Resetting the device because of error", error);
+				force_reset = TRUE;
+			}
```

Comment:
I'm not really sure here - ERR_UNKNOWN can also easily happen when Gammu doesn't understand some response and it could easily lead to reset loop.

I'd rather recommend using [ResetFrequency](https://wammu.eu/docs/manual/smsd/config.html#option-ResetFrequency) to reset phone in defined intervals (eg. weekly).

Reply:
I agree. However, I did not understand what happened until I thoroughly sniffed the AT command stream, and dug in gammu's code. At least, do you think the log files could be more explicit and give hints about this ResetFrequency when UNKNOWN errors are encountered?
I never payed attention to this feature until I saw it in the code (and I swear I read the documentation before using smsd :-) )

I get your point about a confusion about ERR_UNKNWON when Gammu doesn't understand some response. Do you think we should then distinguish two types of unknown errors, i.e. errors when Gammu does not understand, and errors when the dongle actually sends ERROR 500 messages over the wire? Because such an error 500 (= "unknown error" according to the AT standard) seems to be meaningful, at least for my dongle (more than if Gammu did not understand something).
Maybe we should also add a counter to prevent getting in a reboot loop?

My point is, I agree this may be dangerous, I agree I am using a crappy dongle, but I'm quite sure I'm not the only one to get such errors.
And I'm pretty sure 90% of the people getting such errors won't search as deep as I did. Giving them an out-of-the-box solution may ease their lives?

Reply:
The problem is that error 500 is the most frequent you get from the AT modem. So assuming it is good reason to reset doesn't sound reasonable. I can as well mean that Gammu constructed command that modem doesn't understand and resetting in that case would not make the situation any better.